index.rst

```diff
@@ -48,14 +48,126 @@
+
+
+Reducing Registry Access
+========================
+
+When running at scale the datastore reads and writes involving S3 or Google Cloud Storage are robust and load-balancing is handled natively by the cloud infrastructure.
+With a single PostgreSQL registry where all jobs must see an up-to-date view of the registry, reducing connections to the database, the number of transactions running inside the database, and the length of those transactions is critical when attempting to run with hundreds of thousands of simultaneous jobs.
+
+There are many different points at which we can decide to do the registry updates and they are described below.
+
+Immediate
+---------
+
+The current approach used by the ``pipetask`` executor is to call ``butler.get()`` sequentially for each of the requested input datasets (sometimes this read can be deferred, or example when coadding many input images).
+When the ``PipelineTask`` completes each of the output datasets is then stored in the butler with a ``butler.put()``.
+
+It will use the Butler it has been configured to use and in the batch system that is a shared PostgreSQL registry and S3-like object store.
+
+One problem with this approach is that the Butler is designed such that the registry changes and the file writing must all be completed within a database transaction.
+This can result in the database transactions taking a relatively long time for large data files (which first have to be written to local scratch space and then uploaded to the object store).
+Long-lived transactions can cause trouble with database responsiveness when tens of thousands of connections are active.
+
+There are also concerns, as expressed in :jira:`DM-26302`, that the connection management in Butler registry is non-optimal since it disables connection pooling in SQLAlchemy.
+
+.. note::
+  How long does the ``pipetask`` executor keep each registry connection open?
+  Is a single connection used for the duration of execution of each quanta?
+  Or is there one connection used for all the gets at the start and a new connection used for all the puts at the end?
+
+Job-level Pooling
+-----------------
+
+One approach to minimizing registry interactions is to use a local registry during the pipeline execution job and then to batch the registry writes once the pipeline(s) associated with that job finishes.
+
+In this scenario the job would start by exporting the relevant subset of the global registry to a local file (in YAML or JSON format) and importing it into a newly-created local SQLite-based Butler.
+This SQLite database would still refer to the datasets in their original datastore location.
+The pipeline would then be executed using the local registry and would interact as before with the cloud object store.
+When the processing completes the newly-created records would be exported from the local registry and imported to the cloud registry.
```

Comment:
Does "processing completes" include both success and failure?  On failure, I assume having whatever the job produced accessible for debugging is useful.  The question is whether we want to do the merge.

Reply:
Isn't that a function of the workflow management system? If jobs fail doesn't is automatically retain all the local files? Is the job responsible for copying files off the node to a debugging area?

Reply:
The question is whether on failure the job (or run) registry merge is supposed to happen or whether the users will access any outputs separately via the various job registries in the debugging area.  Basically what needs to be the case for when users try to debug a run under normal science/pipeline failures?  A single run repo separate from main repo?  Everything merged back into the main repo?  Individual job repos (which may or may not have the full run info without an explicit merge depending upon how that decision goes)?

Reply:
It seems to me like it's better not to update the central registry if a job fails. If the sqlite file and logs are retained by the workflow managements system that seems to be sufficient for debugging.